presto-cli/src/main/java/com/facebook/presto/cli/QueryPreprocessor.java

```diff
@@ -155,7 +167,9 @@ private static String preprocessQueryInternal(Optional<String> catalog, Optional
 
             // check we got a valid exit code
             if (exitCode != 0) {
-                throw new QueryPreprocessorException("Query preprocessor failed:\n\n" + result);
+                Optional<String> errorMessage = tryGetFutureValue(readStderr, 100, MILLISECONDS)
+                        .flatMap(value -> Optional.ofNullable(emptyToNull(value.trim())));
+                throw new QueryPreprocessorException("Query preprocessor failed: " + errorMessage.orElseGet(() -> "Exit code " + exitCode));
```

Comment:
* exit code might be useful regardless of error message (e.g. `"Query preprocessor exited with %d: %s"`)
* stderr might make produced exception's message hard to read, e.g. when it's java program and stderr contains a stacktrace; or some other multi-line output. Consider `stderr.replace("\n", "\n\t\t| ")` or something.